pkg/controllers/deploymentreplicassyncer/deployment_replicas_syncer_controller.go

```diff
@@ -98,69 +106,112 @@ func (r *DeploymentReplicasSyncer) Reconcile(ctx context.Context, req controller
+		deployment.Name, oldReplicas, deployment.Status.Replicas)
+
+	return controllerruntime.Result{}, nil
+}
 
+// isDeploymentStatusCollected judge whether deployment modification in spec has taken effect and its status has been collected.
+func isDeploymentStatusCollected(deployment *appsv1.Deployment, binding *workv1alpha2.ResourceBinding) bool {
 	// make sure the replicas change in deployment.spec can sync to binding.spec, otherwise retry
 	if deployment.Spec.Replicas == nil || *deployment.Spec.Replicas != binding.Spec.Replicas {
 		klog.V(4).Infof("wait until replicas of binding (%d) equal to replicas of deployment (%d)",
 			binding.Spec.Replicas, *deployment.Spec.Replicas)
-		return controllerruntime.Result{}, fmt.Errorf("retry to wait replicas change sync to binding")
+		return false
 	}
 
 	// make sure the scheduler observed generation equal to generation in binding, otherwise retry
 	if binding.Generation != binding.Status.SchedulerObservedGeneration {
 		klog.V(4).Infof("wait until scheduler observed generation (%d) equal to generation in binding (%d)",
 			binding.Status.SchedulerObservedGeneration, binding.Generation)
-		return controllerruntime.Result{}, fmt.Errorf("retry to wait scheduler observed generation")
+		return false
 	}
 
 	if len(binding.Status.AggregatedStatus) != len(binding.Spec.Clusters) {
 		klog.V(4).Infof("wait until all clusters status collected, got: %d, expected: %d",
 			len(binding.Status.AggregatedStatus), len(binding.Spec.Clusters))
-		return controllerruntime.Result{}, fmt.Errorf("retry to wait status in binding collected")
+		return false
 	}
+
+	bindingStatusReplicas := int32(0)
 	for _, status := range binding.Status.AggregatedStatus {
 		if status.Status == nil {
 			klog.V(4).Infof("wait until aggregated status of cluster %s collected", status.ClusterName)
-			return controllerruntime.Result{}, fmt.Errorf("retry to wait status in binding collected")
+			return false
+		}
+		itemStatus := &appsv1.DeploymentStatus{}
+		if err := json.Unmarshal(status.Status.Raw, itemStatus); err != nil {
+			klog.Errorf("unmarshal status.raw of cluster %s in binding failed: %+v", status.ClusterName, err)
+			return false
 		}
+		if itemStatus.Replicas <= 0 {
+			klog.V(4).Infof("wait until aggregated status replicas of cluster %s collected", status.ClusterName)
+			return false
+		}
+		bindingStatusReplicas += itemStatus.Replicas
 	}
 
-	// update replicas
-	oldReplicas := *deployment.Spec.Replicas
-	deployment.Spec.Replicas = &deployment.Status.Replicas
-	if err := r.Client.Update(ctx, deployment); err != nil {
-		klog.Errorf("failed to update deployment (%s/%s) replicas: %+v", deployment.Namespace, deployment.Name, err)
-		return controllerruntime.Result{}, err
+	if deployment.Status.Replicas != bindingStatusReplicas {
+		klog.V(4).Infof("wait until deployment status replicas (%d) equal to binding status replicas (%d)",
+			deployment.Status.Replicas, bindingStatusReplicas)
+		return false
```

Comment:
Notes: Actually, only these lines has been changed, I added two wait condition.

Other line's change is just because I wrapped a inner retry function  `wait.PollUntilContextTimeout`, for the previous retry interval of `Reconcile` would getting larger and larger, it better to keep same interval in this case.